[ADD] web widget bootstrap toggle

aimsystems opened the pull request:
Added the web_toggle directory which when installed allows for the use of a bootstrap toggle widget within odoo's backend. It is meant to be used for booleans to replace checkboxes with more attractive Yes/No or On/Off toggle's and other options as well.

aimsystems committed:
Oct6_2015_01

petrus-v commented:
That's shame, I've done quite the same taff against v7 #236

aimsystems commented:
Well I havent looked at your V7 but its great really if all of us can expect to find compatible widgets for all of the odoo versions. So if your V7 widget is adopted and my V8 (maybe 9) is accepted then we have covered all the major odoo versions.  I still need to take the time to determine how to overcome the Travis CI build failure which I know nothing about as of now.

legalsylvain commented on web_toggle/__openerp__.py:
```diff
@@ -0,0 +1,17 @@
+    'category': 'Tools',
+    'description':
+        """
+AIM Checkbox Toggle Widget
```
please remove AIM reference in name and description.

legalsylvain commented:
@petrus-v  & @aimsystems : It should be great to name both modules with the same name, and to set the same type of call. (widget="boolean_switch" ...)
So, migration of modules using this module from 7.0 to 8.0 will not require extra work.

@petrus-v : read [ML](https://www.odoo.com/fr_FR/groups/community-59/community-13769111?mode=thread) !

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update bs_toggle.js

aimsystems committed:
Update bs_toggle.xml

aimsystems commented:
I have changed the widget to be enabled using widget="boolean_switch"

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

dreispt commented on web_toggle/__openerp__.py:
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+{
+    'name': 'Bootstrap Toggle Widget',
+    'version': '1.0',
```
`8.0.1.0.0`

dreispt commented on web_toggle/__openerp__.py:
```diff
@@ -0,0 +1,18 @@
+    'version': '1.0',
+    'summary': 'Odoo Checkbox Widget',
+    'category': 'Tools',
+    'description':
```
Missing author and license keys

dreispt replied:
Make the description a README.rst file (remember to use the OCA template)

dreispt commented on web_toggle/static/src/js/bootstrap-toggle.js:
```diff
@@ -0,0 +1,171 @@
+ +function ($) {
```
Does this have lib have a license file to attach to it?

aimsystems replied:
After reviewing the zip file that bootstrap gives you as a download I found a LICENCE file. I have added it next to the to the two files I use from the bootstrap-toggle library.

petrus-v replied:
Hi!

Could you move third party library under /static/src/lib directory ?

aimsystems committed:
Create LICENCE

This is the licence found in the zip file bootstrap-toggle gives you. I hope this is sufficient.

aimsystems committed:
Create LICENCE

aimsystems committed:
Update __openerp__.py

In response to Daniel's comments I have modified __openerp__.py , I apologise for a lack of adherence to the guidelines. I am new to using GITHUB and contributing to OCA and so I am trying to get a handle on all of these details. Thank you for your advice and patience.

dreispt commented on web_toggle/__openerp__.py:
```diff
@@ -0,0 +1,43 @@
+    'website': 'https://github.com/OCA/web',
+    'summary': 'Odoo Checkbox Toggle Widget',
+    'category': 'Tools',
+    'description':
```
Can you remove the description from here and add an OCA README.rst including it's content?

aimsystems replied:
I have removed the description. Waiting for the build to run its checks. Thanks for guiding me through this Daniel.

aimsystems committed:
Create README.rst

Daniel I saw that you suggested I use this README.rst file. I pretty well just copied one I found here in OCA/web and made some modifications. 

Please let me know if this will do. 

Thanks again for you input!

aimsystems committed:
Update __openerp__.py

Removed description

dreispt commented on web_toggle/README.rst:
```diff
@@ -0,0 +1,66 @@
+    :alt: License
+
+Boolean Bootstrap Toggle
+===================
```
Fix this, or rst to html conversion might fail:

```
Boolean Bootstrap Toggle
========================
```

dreispt commented on web_toggle/README.rst:
```diff
@@ -0,0 +1,66 @@
+                }"/>
+
+The above example would create a mini size swich which displayed a 'Yes' when true and 'No' when false would 
+display bootstrap's primary class colour when true and the danger class colour when false.
```
Great!

aimsystems committed:
Update README.rst

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update README.rst

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems commented:
I have updated my project. The tests seem to have all passed. I am not sure what the protocol is to initiate the next stage of acceptance. Any input would be appreciated.

dreispt commented:
Next step is simply to ask for review again, as you did.

dreispt commented on web_toggle/README.rst:
```diff
@@ -0,0 +1,57 @@
+utilized by providing an options dictionary. Go to bootstrap-toggle documentation to learn about all available 
+options such as colours, size, text, shapes, etc. Here is a full example usage.
+
+<field name="my_boolean" widget='boolean_switch' options="{'data-on':'Yes','data-off':'No','data-onstyle':'primary','data-offstyle':'danger','data-size':'mini'}"/>
```
It would be nice to format this [as code](http://sphinx-doc.org/rest.html#source-code), but I won't pick on that.

dreispt commented:
:+1:

aimsystems commented:
Thanks a lot Daniel. I am new to this so I am not sure what the etiquette is regarding turn around time on project stages. I know that we all have a lot of work to do. I do not mind updating the formatting to represent code. I will do so tonight. Just finishing up a little something for work and I will get to it tonight. Thanks again!

dreispt commented:
We're all volunteers, no mandatory times here. You return to it when you can, and reviewers look at it again when they can - if not the former ones, maybe new ones. It's not that rigid.

aimsystems committed:
Update README.rst

aimsystems committed:
Update README.rst

aimsystems commented:
Have updated the README.rst to format the code example as such. Whenever someone has an opportunity to review again I would appreciate it. 

Thanks again!

aimsystems commented:
Hi OCA, I have completed updating my branch. How do I have my widget added to the OCA/web project?

I apologize but this is my first contribution and I am learning as I go here. 

Thanks for your input!

dreispt commented:
You already have my :+1:
Another review is needed.

dreispt commented:
@aimsystems the thing is, anyone can do it not many people contribute with reviews.

legalsylvain commented:
@aimsystems : thanks for your contribution. You can discover OCA coding rules [here](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md) and OCA sharing rules [here](https://odoo-community.org/page/code). specially the section "Get your code merged".

@petrus-v : could you take a look and give your PoV on this PR, as you worked on similar module in V7.

Kind regards.

pedrobaeza commented on web_toggle/README.rst:
```diff
@@ -0,0 +1,60 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License
+
```
Please add here the target link for the badge (according latest template), an upper line for highlighting the title, and other new sections like Try me on runbot. You can see the full template here: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

pedrobaeza commented on web_toggle/static/src/css/LICENCE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Is MIT license compatible with AGPL? I think so, but I want to be sure

pedrobaeza commented:
Thanks for the contribution. I hope this is the first of plenty ones!

I have reviewed the PR, and I would like you to improve these things:
- Change the name to _web_widget_bootstrap_toggle_, because all the widgets addons have this prefix in the name. Including also bootstrap word, you help to specify which king of toggle it is.
- You should mention in any place from which bootstrap version you have extracted the toggle button. Odoo already ships a bootstrap version. Isn't included in this bundle?
- You can provide a view as demo data that changes one existing view on base module that switch the widget for a boolean field. This way, it's not needed to modify nothing to try your module.

aimsystems commented:
Thanks guys. I Will implement your suggestions over the weekend. Thanks again for the prompt feedback. 

Much appreciated.

petrus-v commented on web_toggle/static/src/js/bs_toggle.js:
```diff
@@ -0,0 +1,74 @@
+openerp.web_toggle = function(instance, local) {
+    var _t = instance.web._t,
+        _lt = instance.web._lt;
```
I haven't seen where you are using those variables. If you don't need it, you can remove it;)

aimsystems commented:
Yes, I will follow up on all suggestions over the weekend. 

Thanks again for you support!

petrus-v commented on web_toggle/static/src/js/bs_toggle.js:
```diff
@@ -0,0 +1,74 @@
+    template: 'FieldBooleanToggle',
+    start: function() {
+        var self = this;
+        this.$checkbox = $("input", this.$el);
```
I didn't know that syntax `$("input", this.$el)`, I usually use `this.$el.find("input")`. I won't complain about I've tested it both works!

petrus-v commented:
I've tested on runbot uning chromium (Version 46.0.2490.71 Built on 8.2, running on Debian 8.2 (64-bit)) on res.users form with this setting on active field:

```
<field name="active" widget="boolean_switch"/>
```

I got 3 issues
- In edit mode, when clicking the widget the value doesn't change
- When I leave edit mode discarding change, the widget is just a blue
  rectangle (loosing text and button)
- When I set a user inactive a change record using arrows on top of form view,
  the widget styled still showing the value from the first user.

Do not hesitate to have look what I've done in v7, this king of things haven't change a lot!

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update toggle_widget.xml

aimsystems committed:
UPDATE widget title and add demo user toggle example

aimsystems committed:
Merge branch '8.0' of github.com:aimsystems/web into 8.0

aimsystems committed:
Updating with README and LICENCE

aimsystems committed:
Update users.xml

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Delete __init__.py

aimsystems committed:
Update

aimsystems committed:
Merge branch '8.0' of github.com:aimsystems/web into 8.0

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update bs_toggle.xml

aimsystems committed:
Update toggle_widget.xml

aimsystems committed:
Update

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update toggle_widget.xml

aimsystems committed:
Update

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Merge branch '8.0' of github.com:aimsystems/web into 8.0

aimsystems committed:
Merge branch '8.0' of github.com:aimsystems/web into 8.0

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update bs_toggle.xml

aimsystems committed:
Create __init__.py

aimsystems committed:
Trying to fix

aimsystems committed:
Merge branch '8.0' of github.com:aimsystems/web into 8.0

aimsystems committed:
Delete __init__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update

aimsystems committed:
Merge branch '8.0' of github.com:aimsystems/web into 8.0

aimsystems committed:
Delete __init__.py

aimsystems commented:
Hey guys I have been trying to implement the suggestions. Now my build keeps complaining about the fact that there is no module named web_widget_bootstrap_toggle. Would anyone be able to take a peak and see what it is that I need to change. Thanks

aimsystems commented:
Also thanks petrus-v I have now implemented an example for the active field in users. Also found why it was not working. In my test application I was adding the boostrap-toggle.css and bootstrap-toggle.js which I have now added to the backend.

pedrobaeza commented:
You have forgotten to include file **init**.py in the new module layout. Please also take the occasion to squash all the commits to one, as this is the PR that adds the module.

aimsystems committed:
Create __init__.py

aimsystems commented:
Thanks, how do I squash the commits?

pedrobaeza commented:
You can do it with `git rebase -i --autosquash efb28067065b0a374422d1f68c13f080fc0203ad`

aimsystems commented:
git rebase -i --autosquash efb28067065b0a374422d1f68c13f080fc0203ad
error: could not apply 2a5ec4e... Update toggle_widget.xml

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 2a5ec4e716b5005d2465754fea89c9979de8f679... Update toggle_widget.xml

aimsystems commented:
Git is complaining. Im a real newb with git

aimsystems committed:
Update __init__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __openerp__.py

aimsystems committed:
Update __init__.py

aimsystems committed:
Update users.xml

aimsystems committed:
Update __init__.py

aimsystems committed:
Update __init__.py

aimsystems committed:
Update README.rst

aimsystems committed:
Update users.xml

pedrobaeza commented:
I see. You have a total screwed branch. It's better to copy your module directory, clone again main branch, insert your module in one commit, and then upload it to this branch with a force push (`git push -f`)

aimsystems commented:
Thanks. I need to take some time to review the documentation on proper GIT protocol and the OCA standards and best practices. I really just tried to jump into it. Which may have been a little overzealous.

pedrobaeza commented:
I know it can be hard at the beginning. If you don't get it, just update this PR with latest changes, and I'll fix it up when merging.

dreispt commented:
@aimsystems If you didn't `git push`, the server stored branch is intact. Assuming that `origin` points at your fork you can `git reset --hard origin/8.0`.

aimsystems committed:
Update

aimsystems committed:
Update README.rst

aimsystems commented:
Thanks Daniel and Pedro.I will look through the documentation and get a better handle on GIT and the OCA standards.  Pedro, like you said I completely butchered my branch.

pedrobaeza commented:
@aimsystems, any news on this? Are you going to finish the work?

aimsystems commented:
Hi Pedro, you are right. I probably should see it through. I put it on 
the back burner as I had a lot on my plate. I will take a look at it 
next week and see what needs to be done.

On 2016-10-20 09:01 AM, Pedro M. Baeza wrote:

> @aimsystems https://github.com/aimsystems, any news on this? Are you 
> going to finish the work?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> https://github.com/OCA/web/pull/235#issuecomment-255099179, or mute 
> the thread 
> https://github.com/notifications/unsubscribe-auth/ANnvpTCEnF12AN13EGfGQwfenmEWyjt3ks5q12ZBgaJpZM4GKGYn.
